src/fsharp/FSharp.Core/control.fs

```diff
@@ -1541,8 +1541,8 @@ namespace Microsoft.FSharp.Control
             let continuation (completedTask : Task<_>) : unit =
                 args.aux.trampolineHolder.Protect((fun () ->
                     if completedTask.IsCanceled then
-                        if useCcontForTaskCancellation then args.aux.ccont(new OperationCanceledException())
-                        else args.aux.econt (ExceptionDispatchInfo.Capture(new TaskCanceledException()))
+                        if useCcontForTaskCancellation then args.aux.ccont(new OperationCanceledException(args.aux.token))
```

Comment:
@saul @dsyme this is incorrect. You are passing the async cancellation token which may not be related to the cancellation of the task we're binding on.